Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import json

Read the file in json format and convert it to dataframe

In [ ]:

with open("job_lists.json", "r") as file:
    jobs = json.load(file)
    dataset = pd.DataFrame.from_dict(jobs)

dataset.head()

job_title  \
0                      Remote Data Entry   
1  Data Analyst Intern (Human Resources)   
2                      Jr Data Scientist   
3                           Data Analyst   
4                  Data Scientist Intern   

                                     job_description   seniority_level  \
0  Company Overview\nFormapelec is a dynamic and ...       Entry level   
1  Job Description\nData Analyst Intern\nThe Iowa...        Internship   
2  Company Description\nWho We Are:\nSpring Ventu...  Mid-Senior level   
3  Our client is seeking a skilled Data Analyst t...         Associate   
4  Cardiosense is a digital health company that o...    Not Applicable   

  employment_type                         salary  \
0       Full-time  $40,000.00/yr - $54,000.00/yr   
1      Internship                           None   
2       Full-time                           None   
3       Part-time                           None   
4      Internship                           None   

                       company_name         location   time_posted  \
0                        FORMAPELEC    United States   9 hours ago   
1  State of Iowa - Executive Branch         Ames, IA          None   
2              Spring Venture Group  Kansas City, MO  20 hours ago   
3                   Stealth Startup    United States          None   
4                       Cardiosense      Chicago, IL          None   

  num_applicants                                           job_link  
0           None  https://www.linkedin.com/jobs/view/remote-data...  
1  70 applicants  https://www.linkedin.com/jobs/view/data-analys...  
2           None  https://www.linkedin.com/jobs/view/jr-data-sci...  
3           None  https://www.linkedin.com/jobs/view/data-analys...  
4           None  https://www.linkedin.com/jobs/view/data-scient...

In [ ]:
len(dataset)

8938

In [ ]:
dataset.isnull().sum()

job_title          1856
job_description    1856
seniority_level    1856
employment_type    1856
salary             5489
company_name       1856
location           1856
time_posted        3832
num_applicants     5953
job_link           1856
dtype: int64

The dataset contains a significant number of null values in various columns. However, rows with missing values in the job_title column (1856 entries) are being removed to ensure the integrity of the analysis.

In [ ]:
dataset = dataset.dropna(subset="job_title")

In [ ]:
dataset.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8933     True
8934     True
8935     True
8936     True
8937     True
Length: 7082, dtype: bool

Since there seems to be the duplicate values they all needs to be removed.

In [ ]:
dataset.drop_duplicates(inplace=True)

In [ ]:
len(dataset)

1178

After dropping the duplicates there is only 1178 entries exists it means there were too many duplicated values.

In [ ]:
dataset["job_title"].unique()

array(['Remote Data Entry', 'Data Analyst Intern (Human Resources)',
       'Jr Data Scientist', 'Data Analyst', 'Data Scientist Intern',
       'Agency Data Analyst',
       "Analytics/Data Science Master's Level Internship (Summer 2025 - Remote)",
       'Data Analytics Intern',
       'Graduate Student Data Scientist - Intern',
       'Fully Remote Data Entry Clerk / Work From Home Typist',
       'Research & Analytics Intern, Summer 2025',
       'Data Analyst - Fully Remote in the US', 'Product Data Scientist',
       'Staff Data Analyst, Credit Analytics',
       "Don't see an opening that matches your background? Join our talent community here!",
       'Remote Data Entry Clerk / Typing - Part Time or Full Time',
       'Data Entry - Remote (WFH)',
       'Data Scientist I - User Growth Analytics (LatAm)',
       'Virtual Data Entry Clerk', 'Digital Marketing Manager',
       'Business Analyst, Marketing, Amazon Key', 'Data Analyst I',
       'Data analyst with Python', 'Growth 

Remove inappropriate job title from the dataset

In [ ]:
dataset = dataset.drop(dataset[dataset["job_title"] == "Don't see an opening that matches your background? Join our talent community here!"].index)

In [ ]:
results = dataset[dataset["job_title"].str.contains("data", case=False, na=False)]

array(['Remote Data Entry', 'Data Analyst Intern (Human Resources)',
       'Jr Data Scientist', 'Data Analyst', 'Data Scientist Intern',
       'Agency Data Analyst',
       "Analytics/Data Science Master's Level Internship (Summer 2025 - Remote)",
       'Data Analytics Intern',
       'Graduate Student Data Scientist - Intern',
       'Fully Remote Data Entry Clerk / Work From Home Typist',
       'Data Analyst - Fully Remote in the US', 'Product Data Scientist',
       'Staff Data Analyst, Credit Analytics',
       'Remote Data Entry Clerk / Typing - Part Time or Full Time',
       'Data Entry - Remote (WFH)',
       'Data Scientist I - User Growth Analytics (LatAm)',
       'Virtual Data Entry Clerk', 'Data Analyst I',
       'Data analyst with Python', 'Data and Insights Analyst',
       'Data and Visualization Analyst', 'Entry Level Data Analyst',
       'Business Data Analyst', 'Data Entry Clerk',
       'Data Entry Specialist - Tempe', 'Data Engineer Intern',
       'Data Entr

The focus is on roles related to Data Analyst, Data Scientist, Data Engineer, Data Analytics, and Data Entry. Job titles containing these keywords (with variations) in the job_title column have been standardized to their respective roles to maintain consistency in the analysis.

In [ ]:
dataset.loc[dataset["job_title"].str.contains("data analyst", case=False, na=False), "job_title"]="Data Analyst"
dataset.loc[dataset["job_title"].str.contains(r"data\s*(scientist|science)", case=False, na=False), "job_title"]="Data Scientist"
dataset.loc[dataset["job_title"].str.contains("data engineer", case=False, na=False), "job_title"]="Data Engineer"
dataset.loc[dataset["job_title"].str.contains("data analytics", case=False, na=False), "job_title"]="Data Analytics"
dataset.loc[dataset["job_title"].str.contains("data entry", case=False, na=False), "job_title"]="Data Entry"


/var/folders/nl/mq6lrx555337t5yq69ftc0jr0000gn/T/ipykernel_20627/1007938783.py:2: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  dataset.loc[dataset["job_title"].str.contains(r"data\s*(scientist|science)", case=False, na=False), "job_title"]="Data Scientist"


The salary data was initially in a range format. For the analysis the values were split into two separate columns: min_salary and max_salary.

In [ ]:
salary = dataset["salary"].str.split("-", expand=True)
dataset["min_salary"] = salary[0]
dataset["max_salary"] = salary[1]

In [ ]:
dataset = dataset.drop("salary", axis=1)

In [ ]:
dataset.head()

job_title                                    job_description  \
0      Data Entry  Company Overview\nFormapelec is a dynamic and ...   
1    Data Analyst  Job Description\nData Analyst Intern\nThe Iowa...   
2  Data Scientist  Company Description\nWho We Are:\nSpring Ventu...   
3    Data Analyst  Our client is seeking a skilled Data Analyst t...   
4  Data Scientist  Cardiosense is a digital health company that o...   

    seniority_level employment_type                      company_name  \
0       Entry level       Full-time                        FORMAPELEC   
1        Internship      Internship  State of Iowa - Executive Branch   
2  Mid-Senior level       Full-time              Spring Venture Group   
3         Associate       Part-time                   Stealth Startup   
4    Not Applicable      Internship                       Cardiosense   

          location   time_posted num_applicants  \
0    United States   9 hours ago           None   
1         Ames, IA          None  70 applicants   
2  Kansas City, MO  20 hours ago           None   
3    United States          None           None   
4      Chicago, IL          None           None   

                                            job_link      min_salary  \
0  https://www.linkedin.com/jobs/view/remote-data...  $40,000.00/yr    
1  https://www.linkedin.com/jobs/view/data-analys...            None   
2  https://www.linkedin.com/jobs/view/jr-data-sci...            None   
3  https://www.linkedin.com/jobs/view/data-analys...            None   
4  https://www.linkedin.com/jobs/view/data-scient...            None   

       max_salary  
0   $54,000.00/yr  
1            None  
2            None  
3            None  
4            None

The salary column contains both annual and hourly rates. To standardize the values, hourly rates were converted to annual salaries by multiplying the hourly rate by 40 hours per week and 52 weeks per year. The conversion was applied to both min_salary and max_salary columns.

In [ ]:
def convert_hourly_salary_to_anuual(salary):
    if salary:
        salary = salary.replace("$", "")
        if "hr" in salary:
            hourly_rate = salary.split("/")[0]
            yearly_salary = float(hourly_rate) * 40 * 52
            return yearly_salary
        elif "yr" in salary:
            annual_salary = salary.split("/")[0]
            return annual_salary
    else:
        return None

dataset["min_salary"] = dataset["min_salary"].apply(convert_hourly_salary_to_anuual)
dataset["max_salary"] = dataset["max_salary"].apply(convert_hourly_salary_to_anuual)


In [ ]:
dataset.head()

job_title                                    job_description  \
0      Data Entry  Company Overview\nFormapelec is a dynamic and ...   
1    Data Analyst  Job Description\nData Analyst Intern\nThe Iowa...   
2  Data Scientist  Company Description\nWho We Are:\nSpring Ventu...   
3    Data Analyst  Our client is seeking a skilled Data Analyst t...   
4  Data Scientist  Cardiosense is a digital health company that o...   

    seniority_level employment_type                      company_name  \
0       Entry level       Full-time                        FORMAPELEC   
1        Internship      Internship  State of Iowa - Executive Branch   
2  Mid-Senior level       Full-time              Spring Venture Group   
3         Associate       Part-time                   Stealth Startup   
4    Not Applicable      Internship                       Cardiosense   

          location   time_posted num_applicants  \
0    United States   9 hours ago           None   
1         Ames, IA          None  70 applicants   
2  Kansas City, MO  20 hours ago           None   
3    United States          None           None   
4      Chicago, IL          None           None   

                                            job_link min_salary  max_salary  
0  https://www.linkedin.com/jobs/view/remote-data...  40,000.00   54,000.00  
1  https://www.linkedin.com/jobs/view/data-analys...       None        None  
2  https://www.linkedin.com/jobs/view/jr-data-sci...       None        None  
3  https://www.linkedin.com/jobs/view/data-analys...       None        None  
4  https://www.linkedin.com/jobs/view/data-scient...       None        None

In [ ]:
dataset["num_applicants"].unique()

array([None, '70 applicants', '134 applicants', '97 applicants',
       '86 applicants', '124 applicants', '69 applicants',
       '162 applicants', '139 applicants', '199 applicants',
       '125 applicants', '196 applicants', '113 applicants',
       '25 applicants', '74 applicants', '30 applicants', '47 applicants',
       '58 applicants', '132 applicants', '111 applicants',
       '38 applicants', '88 applicants', '48 applicants',
       '118 applicants', '114 applicants', '50 applicants',
       '42 applicants', '193 applicants', '190 applicants',
       '75 applicants', '79 applicants', '73 applicants', '41 applicants',
       '52 applicants', '46 applicants', '159 applicants',
       '40 applicants', '187 applicants', '27 applicants',
       '188 applicants', '67 applicants', '63 applicants',
       '106 applicants', '29 applicants', '34 applicants',
       '103 applicants', '68 applicants', '28 applicants',
       '35 applicants', '189 applicants', '136 applicants',
       '137

In [ ]:
def remove_texts_from_applicants(applicant):
    if applicant:
        applicant = applicant.replace("applicants", "")
        return int(applicant)
    else:
        return 0

dataset["num_applicants"] = dataset["num_applicants"].apply(remove_texts_from_applicants)

The num_applicants column contains text along with numeric values. To clean the data, the text "applicants" was removed, and the remaining numeric values were converted to integers. Rows without values were set to 0.

In [ ]:
 dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1177 entries, 0 to 8924
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   job_title        1177 non-null   object
 1   job_description  1177 non-null   object
 2   seniority_level  1177 non-null   object
 3   employment_type  1177 non-null   object
 4   company_name     1177 non-null   object
 5   location         1177 non-null   object
 6   time_posted      880 non-null    object
 7   num_applicants   1177 non-null   int64 
 8   job_link         1177 non-null   object
 9   min_salary       575 non-null    object
 10  max_salary       575 non-null    object
dtypes: int64(1), object(10)
memory usage: 110.3+ KB


In [ ]:
dataset["min_salary"] = dataset["min_salary"].str.replace(",", '').astype(float)
dataset["max_salary"] = dataset["max_salary"].str.replace(",", '').astype(float)

In [ ]:
dataset["location"].unique()

array(['United States', 'Ames, IA', 'Kansas City, MO', 'Chicago, IL',
       'Denver, CO', 'Atlanta, GA', 'Salt Lake City, UT',
       'Lake Oswego, OR', 'Scranton, PA', 'Burbank, CA', 'Anaheim, CA',
       'Washington County, OR', 'Cincinnati Metropolitan Area',
       'Durham, NC', 'Houston, TX', 'New York, NY',
       'Omaha Metropolitan Area', 'Fort Stewart, GA', 'Santa Monica, CA',
       'Killeen, TX', 'Dallas, TX', 'San Jose, CA', 'Beverly Hills, CA',
       'Austin, TX', 'Atlanta Metropolitan Area', 'Washington, DC',
       'Hialeah, FL', 'Los Angeles, CA', 'Seattle, WA', 'New Albany, OH',
       'Jefferson City, MO', 'McLean, VA', 'Tempe, AZ', 'Newark, NJ',
       'New York County, NY', 'Boston, MA', 'Irving, TX', 'Tampa, FL',
       'Deerfield, IL', 'Raleigh, NC', 'Austin, Texas Metropolitan Area',
       'Hawthorne, CA', 'Tucson, AZ', 'Bellevue, WA',
       'Colorado, United States', 'Alpharetta, GA', 'Ashburn, VA',
       'Bozeman, MT', 'Denver Metropolitan Area', 'Carbonda

For state based job analysis state_abbr column was created by extracting the state abbreviation from the location column.

In [ ]:
dataset["state_abbr"] = dataset["location"].apply(lambda location: location.split(",")[1].strip() if "," in location else None)

In [ ]:
dataset.head(50)

job_title  \
0                                      Data Entry   
1                                    Data Analyst   
2                                  Data Scientist   
3                                    Data Analyst   
4                                  Data Scientist   
5                                    Data Analyst   
6                                  Data Scientist   
7                                  Data Analytics   
8                                  Data Scientist   
9                                      Data Entry   
19       Research & Analytics Intern, Summer 2025   
24                                   Data Analyst   
38                                 Data Scientist   
39                                   Data Analyst   
45                                     Data Entry   
49                                   Data Analyst   
56                                     Data Entry   
58                                     Data Entry   
69                                 Data Scientist   
74                                     Data Entry   
76                      Digital Marketing Manager   
89                                   Data Analyst   
99                                   Data Analyst   
116       Business Analyst, Marketing, Amazon Key   
117                                  Data Analyst   
118                                  Data Analyst   
119                    Growth Marketing Associate   
129                       Machine Learning Intern   
167                     Data and Insights Analyst   
168       Business Analyst, Marketing, Amazon Key   
176                                  Data Analyst   
178                Data and Visualization Analyst   
189                                  Data Analyst   
216  Summer 2025 Business Intelligence Internship   
217                              UX Researcher II   
229                              Power BI Analyst   
258                                  Data Analyst   
266                                  Data Analyst   
267               Policy and Research Coordinator   
268                                  Data Analyst   
269                                    Data Entry   
278                                  Data Analyst   
279                              Business Analyst   
319                                  Data Analyst   
324       Business Analyst, Marketing, Amazon Key   
325                                    Data Entry   
326                                 Data Engineer   
329                                    Data Entry   
339                                 Data Engineer   
359                     Image Annotation Reviewer   

                                       job_description   seniority_level  \
0    Company Overview\nFormapelec is a dynamic and ...       Entry level   
1    Job Description\nData Analyst Intern\nThe Iowa...        Internship   
2    Company Description\nWho We Are:\nSpring Ventu...  Mid-Senior level   
3    Our client is seeking a skilled Data Analyst t...         Associate   
4    Cardiosense is a digital health company that o...    Not Applicable   
5    The Stellarix Group\nAgency Data Analyst\nRepo...  Mid-Senior level   
6    Req125243\nInternship Overview\nThe Home Depot...    Not Applicable   
7    Department:\nTechnology\nLocation:\nSalt Lake ...        Internship   
8    About Us\nRapta is revolutionizing American ma...        Internship   
9    Job Overview\nJoin our esteemed Learning & Dev...       Entry level   
19   About the Role & Program:\nDisney Entertainmen...  Mid-Senior level   
24   Job Summary\nThe Analytics team at WorkCare pr...       Entry level   
38   Rec Room is the best place to build and play g...       Entry level   
39   About Upstart\nUpstart is a leading AI lending...  Mid-Senior level   
45   Job Overview\nJoin our esteemed Learning & Dev...       Entry level   
49   Job Title: BI Dashboard Analyst - Level 1\nDur...         Associate   
56   Job Overview\nWe are actively seeking a

To convert the abbreviation to states full name the new dataset that contatins the information for this is loaded.

In [ ]:
states = pd.read_csv("States.csv")
states.head()

State Standard Postal
0     Alabama     Ala.     AL
1      Alaska   Alaska     AK
2     Arizona    Ariz.     AZ
3    Arkansas     Ark.     AR
4  California   Calif.     CA

Remove unnecessary columns

In [ ]:
states = states.drop("Standard", axis=1)

Merge the dataset using the matching column which is state abbreviations.

In [ ]:
dataset = dataset.merge(states, how="left", left_on="state_abbr", right_on="Postal")
dataset.head(50)

job_title  \
0                                     Data Entry   
1                                   Data Analyst   
2                                 Data Scientist   
3                                   Data Analyst   
4                                 Data Scientist   
5                                   Data Analyst   
6                                 Data Scientist   
7                                 Data Analytics   
8                                 Data Scientist   
9                                     Data Entry   
10      Research & Analytics Intern, Summer 2025   
11                                  Data Analyst   
12                                Data Scientist   
13                                  Data Analyst   
14                                    Data Entry   
15                                  Data Analyst   
16                                    Data Entry   
17                                    Data Entry   
18                                Data Scientist   
19                                    Data Entry   
20                     Digital Marketing Manager   
21                                  Data Analyst   
22                                  Data Analyst   
23       Business Analyst, Marketing, Amazon Key   
24                                  Data Analyst   
25                                  Data Analyst   
26                    Growth Marketing Associate   
27                       Machine Learning Intern   
28                     Data and Insights Analyst   
29       Business Analyst, Marketing, Amazon Key   
30                                  Data Analyst   
31                Data and Visualization Analyst   
32                                  Data Analyst   
33  Summer 2025 Business Intelligence Internship   
34                              UX Researcher II   
35                              Power BI Analyst   
36                                  Data Analyst   
37                                  Data Analyst   
38               Policy and Research Coordinator   
39                                  Data Analyst   
40                                    Data Entry   
41                                  Data Analyst   
42                              Business Analyst   
43                                  Data Analyst   
44       Business Analyst, Marketing, Amazon Key   
45                                    Data Entry   
46                                 Data Engineer   
47                                    Data Entry   
48                                 Data Engineer   
49                     Image Annotation Reviewer   

                                      job_description   seniority_level  \
0   Company Overview\nFormapelec is a dynamic and ...       Entry level   
1   Job Description\nData Analyst Intern\nThe Iowa...        Internship   
2   Company Description\nWho We Are:\nSpring Ventu...  Mid-Senior level   
3   Our client is seeking a skilled Data Analyst t...         Associate   
4   Cardiosense is a digital health company that o...    Not Applicable   
5   The Stellarix Group\nAgency Data Analyst\nRepo...  Mid-Senior level   
6   Req125243\nInternship Overview\nThe Home Depot...    Not Applicable   
7   Department:\nTechnology\nLocation:\nSalt Lake ...        Internship   
8   About Us\nRapta is revolutionizing American ma...        Internship   
9   Job Overview\nJoin our esteemed Learning & Dev...       Entry level   
10  About the Role & Program:\nDisney Entertainmen...  Mid-Senior level   
11  Job Summary\nThe Analytics team at WorkCare pr...       Entry level   
12  Rec Room is the best place to build and play g...       Entry level   
13  About Upstart\nUpstart is a leading AI lending...  Mid-Senior level   
14  Job Overview\nJoin our esteemed Learning & Dev...       Entry level   
15  Job Title: BI Dashboard Analyst - Level 1\nDur...         Associate   
16  Job Overview\nWe are actively seeking a dilige...       Entry level   
17  Job Title:\nData Entry - Remote 

Finally, export the cleaned dataset into csv format

In [ ]:
dataset.to_csv("clean_dataset.csv", index=False)